andela/ah-legion-backend
# [#164046247] Users can verify their accounts

#### Description
Whenever users create an account, it is good practice to have them verify their email addresses so we can be sure that the users are indeed real and that they own the accounts they used for registration. This feature implements email verification by sending a JWT token to the user email address that should be decoded. If the token is valid, the token is decoded and the user is redirected to a callback URL that they provided when registering.
#### Type of change
- [x] Breaking Change(Users must provide a callback url when registering, Unverified users cannot log in )
- [x] Non-breaking changes ( Users can get a new verification token if the old one expires, Users can verify their accounts via email)

#### How Has This Been Tested?

#### Checklist:
- [x] I have tested all new functions
- [x] Feature passes team flake8 standards
- [x] Refactored prior tests to make them pass considering I have breaking changes

### How can this be manually tested?
- Follow instructions on the README page and set up your environment locally.
- Check out to this branch `ft-email-verification-164046247` and install requirements by running `pip install -r requirements.txt`
- Register a user using a valid email address, and be sure to add a callback URL. It can be any URL of your choice. Here is an example:
 ```
{"user": {
"username": "cray",
"email": "cray@mail.com",
"password": "password",
"callback_url": "http://www.youtube.com"}}
```

- You should get a response that you check your email address for a verification link. Copy and paste that link in your browser and you should be redirected to the callback url you provided. Verified users should not be able to verify again, and users should only be able to verify their own email addresses.
- To get a new verification token, navigate to the endpoint `/api/token/` and pass in your username and email. If you pass in incorrect information, you should not be able to get a reset token. Only registered users can get a reset token. To verifiy the token, simply go to your email address and copy and paste the link in your browser. You should be redirected to the callback URL you provided

#### PT
[#164046247](https://www.pivotaltracker.com/story/show/164046247)
